File: Android/app/src/main/java/com/example/supplementtracker/service/SyncConflictAnalyzer.kt
package com.example.supplementtracker.service

import com.example.supplementtracker.domain.export.OAKBackupDataDTO
import com.example.supplementtracker.domain.export.OAKBackupHistoryDTO
import com.example.supplementtracker.domain.export.OAKBackupSupplementDTO
import com.example.supplementtracker.domain.model.UserSupplement
import com.example.supplementtracker.domain.repository.IntakeRecord
import com.example.supplementtracker.domain.util.DoseEventKey
import java.util.Locale

object SyncConflictAnalyzer {
    fun analyze(
        backup: OAKBackupDataDTO,
        localSupplements: Map<String, UserSupplement>,
        localRecords: Map<String, IntakeRecord>,
        localRecordsByDoseKey: Map<String, IntakeRecord>
    ): SyncConflictPreview {
        var preview = SyncConflictPreview()
        backup.stack.forEach { remote ->
            preview = add(preview, supplementDecision(remote, localSupplements))
        }
        backup.history.forEach { remote ->
            preview = add(preview, historyDecision(remote, localRecords, localRecordsByDoseKey))
        }
        return preview
    }

    private fun supplementDecision(
        remote: OAKBackupSupplementDTO,
        locals: Map<String, UserSupplement>
    ): SyncConflictDecision {
        val local = locals[remote.id.lowercase(Locale.ROOT)] ?: return SyncConflictDecision.NO_CONFLICT
        val localTs = maxOf(local.updatedAtEpochMs, local.deletedAtEpochMs ?: 0L)
        val remoteTs = maxOf(remote.updatedAtEpochMs, remote.deletedAtEpochMs ?: 0L)
        return SyncConflictPolicy.decide(true, localTs, remoteTs, supplementDiffers(remote, local))
    }

    private fun historyDecision(
        remote: OAKBackupHistoryDTO,
        locals: Map<String, IntakeRecord>,
        byDoseKey: Map<String, IntakeRecord>
    ): SyncConflictDecision {
        val normalizedSupplementId = remote.supplementId.lowercase(Locale.ROOT)
        val doseKey = DoseEventKey.make(normalizedSupplementId, remote.dateEpochMs)
        val local = byDoseKey[doseKey] ?: locals[remote.id.lowercase(Locale.ROOT)]
            ?: return SyncConflictDecision.NO_CONFLICT
        val remoteTs = remote.updatedAtEpochMs.takeIf { it > 0L } ?: remote.dateEpochMs
        val differs = local.status != remote.status ||
            local.supplementId.lowercase(Locale.ROOT) != normalizedSupplementId
        return SyncConflictPolicy.decide(true, local.updatedAtEpochMs, remoteTs, differs)
    }

    private fun supplementDiffers(remote: OAKBackupSupplementDTO, local: UserSupplement): Boolean {
        if (remote.deletedAtEpochMs != local.deletedAtEpochMs) return true
        val fields = remote.modifiedFields
        if (applies(fields, "name") && remote.name != local.name) return true
        if (applies(fields, "dailyDose") && remote.dailyDose != local.dailyDose) return true
        if (applies(fields, "intakeTime") && remote.intakeTime != local.intakeTime) return true
        if (applies(fields, "startDate") && remote.startDate != local.startDate.toString()) return true
        if (applies(fields, "lastTakenLocalDate") && remote.lastTakenLocalDate != local.lastTakenLocalDate?.toString()) return true
        return applies(fields, "cycle") && cycleDiffers(remote, local)
    }

    private fun cycleDiffers(remote: OAKBackupSupplementDTO, local: UserSupplement): Boolean {
        val cycle = local.cycleConfig
        return remote.cycle.isContinuous != cycle.isContinuous ||
            remote.cycle.daysOn != cycle.daysOn || remote.cycle.daysOff != cycle.daysOff ||
            remote.cycle.durationMonths != cycle.durationMonths || remote.cycle.intervalDays != cycle.intervalDays ||
            remote.cycle.weeklyWeekdaysMask != cycle.weeklyRecurrence?.weekdaysMask ||
            remote.cycle.weeklyIntervalWeeks != cycle.weeklyRecurrence?.intervalWeeks ||
            remote.cycle.weeklyAnchorDate != cycle.weeklyRecurrence?.anchorDate?.toString()
    }

    private fun applies(fields: Set<String>?, name: String): Boolean = fields == null || fields.contains(name)

    private fun add(preview: SyncConflictPreview, decision: SyncConflictDecision): SyncConflictPreview = when (decision) {
        SyncConflictDecision.REMOTE_WINS -> preview.copy(remoteWins = preview.remoteWins + 1)
        SyncConflictDecision.LOCAL_WINS -> preview.copy(localWins = preview.localWins + 1)
        SyncConflictDecision.TIE_LOCAL_WINS -> preview.copy(tieLocalWins = preview.tieLocalWins + 1)
        SyncConflictDecision.NO_CONFLICT -> preview
    }
}

File: iOS/Services/SyncConflictAnalyzer.swift
import Foundation
import SwiftData

@MainActor
enum SyncConflictAnalyzer {
    static func analyze(
        backup: OAKBackupData,
        clientId: UUID,
        context: ModelContext
    ) throws -> SyncConflictPreview {
        let supplements = try context.fetch(FetchDescriptor<UserSupplement>())
            .filter { $0.client?.id == clientId }
        let records = try context.fetch(FetchDescriptor<IntakeRecord>())
            .filter { $0.supplement?.client?.id == clientId }
        let locals = Dictionary(uniqueKeysWithValues: supplements.map { ($0.id, $0) })
        let byId = Dictionary(uniqueKeysWithValues: records.map { ($0.id, $0) })
        let byDoseKey = bestRecordsByDoseKey(records)
        return analyzeBackup(backup, supplements: locals, records: byId, byDoseKey: byDoseKey)
    }

    private static func analyzeBackup(
        _ backup: OAKBackupData,
        supplements: [UUID: UserSupplement],
        records: [UUID: IntakeRecord],
        byDoseKey: [String: IntakeRecord]
    ) -> SyncConflictPreview {
        var preview = SyncConflictPreview()
        for remote in backup.stack {
            preview = add(preview, decision(remote: remote, supplements: supplements))
        }
        for remote in backup.history {
            preview = add(preview, decision(remote: remote, records: records, byDoseKey: byDoseKey))
        }
        return preview
    }

    private static func decision(
        remote: OAKBackupSupplement,
        supplements: [UUID: UserSupplement]
    ) -> SyncConflictDecision {
        guard let id = UUID(uuidString: remote.id), let local = supplements[id] else { return .noConflict }
        let localTs = max(local.updatedAtEpochMs, local.deletedAtEpochMs ?? 0)
        let remoteTs = max(remote.updatedAtEpochMs, remote.deletedAtEpochMs ?? 0)
        return SyncConflictPolicy.decide(
            localExists: true,
            localTs: localTs,
            remoteTs: remoteTs,
            differs: supplementDiffers(remote, local)
        )
    }

    private static func decision(
        remote: OAKBackupHistory,
        records: [UUID: IntakeRecord],
        byDoseKey: [String: IntakeRecord]
    ) -> SyncConflictDecision {
        guard let supplementId = UUID(uuidString: remote.supplementId) else { return .noConflict }
        let doseKey = DoseEventKey.make(supplementId: supplementId, scheduledAtEpochMs: remote.dateEpochMs)
        let stableId = DoseEventKey.stableUUID(from: doseKey)
        guard let local = byDoseKey[doseKey] ?? records[stableId] else { return .noConflict }
        let remoteTs = remote.updatedAtEpochMs > 0 ? remote.updatedAtEpochMs : remote.dateEpochMs
        let differs = local.status != remote.status || local.supplement?.id != supplementId
        return SyncConflictPolicy.decide(
            localExists: true,
            localTs: local.updatedAtEpochMs,
            remoteTs: remoteTs,
            differs: differs
        )
    }

    private static func supplementDiffers(_ remote: OAKBackupSupplement, _ local: UserSupplement) -> Bool {
        if remote.deletedAtEpochMs != local.deletedAtEpochMs { return true }
        let fields = remote.modifiedFields
        if applies(fields, "name") && remote.name != local.name { return true }
        if applies(fields, "dailyDose") && remote.dailyDose != local.dailyDose { return true }
        if applies(fields, "intakeTime") && remote.intakeTime != local.intakeTime { return true }
        if applies(fields, "startDate") && remote.startDate != LocalDayCodec.string(from: local.startDate) { return true }
        if applies(fields, "lastTakenLocalDate") && remote.lastTakenLocalDate != local.lastTakenLocalDate { return true }
        return applies(fields, "cycle") && cycleDiffers(remote.cycle, local.cycleConfig)
    }

    private static func cycleDiffers(_ remote: SupplementExportCycle, _ local: CycleConfig) -> Bool {
        remote.isContinuous != local.isContinuous || remote.daysOn != local.daysOn ||
            remote.daysOff != local.daysOff || remote.durationMonths != local.durationMonths ||
            remote.intervalDays != local.intervalDays ||
            remote.weeklyWeekdaysMask != local.weeklyRecurrence?.weekdaysMask ||
            remote.weeklyIntervalWeeks != local.weeklyRecurrence?.intervalWeeks ||
            remote.weeklyAnchorDate != local.weeklyRecurrence.map { LocalDayCodec.string(from: $0.anchorDate) }
    }

    private static func applies(_ fields: Set<String>?, _ field: String) -> Bool {
        fields == nil || fields?.contains(field) == true
    }

    private static func bestRecordsByDoseKey(_ records: [IntakeRecord]) -> [String: IntakeRecord] {
        var best: [String: IntakeRecord] = [:]
        for record in records {
            guard let supplementId = record.supplement?.id else { continue }
            let epochMs = Int64(record.date.timeIntervalSince1970 * 1000)
            let key = DoseEventKey.make(supplementId: supplementId, scheduledAtEpochMs: epochMs)
            let currentUpdatedAt = best[key]?.updatedAtEpochMs ?? Int64.min
            if currentUpdatedAt < record.updatedAtEpochMs { best[key] = record }
        }
        return best
    }

    private static func add(_ preview: SyncConflictPreview, _ decision: SyncConflictDecision) -> SyncConflictPreview {
        var updated = preview
        switch decision {
        case .remoteWins: updated.remoteWins += 1
        case .localWins: updated.localWins += 1
        case .tieLocalWins: updated.tieLocalWins += 1
        case .noConflict: break
        }
        return updated
    }

}
